drv/lpc55-update-server/src/main.rs

```diff
@@ -402,6 +387,132 @@ impl idl::InOrderUpdateImpl for ServerImpl<'_> {
+        )?;
+        indirect_flash_read_words(
+            &mut self.flash,
+            CFPA_PONG_FLASH_WORD,
+            core::slice::from_mut(&mut pong_header),
+        )?;
+
+        // Work out where to read the authoritative contents from.
+        let val = if ping_header[1] >= pong_header[1] {
+            (CFPA_PING_FLASH_WORD, ping_header[1])
+        } else {
+            (CFPA_PONG_FLASH_WORD, pong_header[1])
+        };
+
+        Ok(val)
+    }
+
+    // Return the persistent and transient boot preferences
+    fn boot_preferences(
+        &mut self,
+    ) -> Result<(SlotId, Option<SlotId>, Option<SlotId>), UpdateError> {
+        let (cfpa_word_number, cfpa_version) =
+            self.cfpa_word_number_and_version()?;
+
+        // Read the authoritative boot selection
+        let boot_selection_word_number = cfpa_word_number + 0x10;
+        let mut boot_selection_word = [0u32; 4];
+        indirect_flash_read_words(
+            &mut self.flash,
+            boot_selection_word_number,
+            core::slice::from_mut(&mut boot_selection_word),
+        )?;
+
+        // Read the scratch boot version
+        let mut scratch_header = [0u32; 4];
+        indirect_flash_read_words(
+            &mut self.flash,
+            CFPA_SCRATCH_FLASH_WORD,
+            core::slice::from_mut(&mut scratch_header),
+        )?;
+
+        // We only have a pending preference if the scratch CFPA page is newer
+        // than the authoritative page.
+        let pending_persistent_boot_preference =
+            if scratch_header[1] > cfpa_version {
+                // Read the scratch boot selection
+                let scratch_boot_selection_word_number =
+                    CFPA_SCRATCH_FLASH_WORD + 0x10;
+                let mut scratch_boot_selection_word = [0u32; 4];
+                indirect_flash_read_words(
+                    &mut self.flash,
+                    scratch_boot_selection_word_number,
+                    core::slice::from_mut(&mut scratch_boot_selection_word),
+                )?;
+                Some(boot_preference_from_flash_word(
+                    &scratch_boot_selection_word,
+                ))
+            } else {
+                None
+            };
+
+        // Check the authoritative persistent boot selection bit
+        let persistent_boot_preference =
+            boot_preference_from_flash_word(&boot_selection_word);
+
+        // We only support persistent override at this point
+        // We need to read the magic ram value to fill this in.
+        let transient_boot_preference = None;
+
+        Ok((
+            persistent_boot_preference,
+            pending_persistent_boot_preference,
+            transient_boot_preference,
+        ))
+    }
+}
+
+// Return the preferred slot to boot from for a given CFPA boot selection
+// flash word.
+fn boot_preference_from_flash_word(flash_word: &[u32; 4]) -> SlotId {
+    if flash_word[0] & 1 == 0 {
```

Comment:
Add a comment here explaining that this matches the logic in bootleby

Reply:
Good idea!

Reply:
done